docs/src/main/asciidoc/security-authorize-web-endpoints-reference.adoc

```diff
@@ -386,6 +386,186 @@ It is possible to use multiple expressions in the role definition.
+import jakarta.ws.rs.QueryParam;
+
+import io.quarkus.security.PermissionsAllowed;
+
+import java.security.Permission;
+import java.util.Collection;
+import java.util.Collections;
+
+@Path("/crud")
+public class CRUDResource {
+
+    @PermissionsAllowed("create") <1>
+    @PermissionsAllowed("update")
+    @POST
+    public String createOrUpdate() {
+        return "modified";
+    }
+
+    @PermissionsAllowed({"see:detail", "see:all", "read"}) <2>
+    @GET
+    @Path("/id/{id}")
+    public String getItem(String id) {
+        return "item-detail-" + id;
+    }
+
+    @PermissionsAllowed(value = "permission-1", permission = CustomPermission.class) <3>
+    @GET
+    public Collection<String> list(@QueryParam("query-options") String queryOptions) {
+        // your business logic comes here
+        return Collections.emptySet();
+    }
+
+    public static class CustomPermission extends Permission {
+
+        public CustomPermission(String name) {
+            super(name);
+        }
+
+        @Override
+        public boolean implies(Permission permission) {
+            var event = Arc.container().instance(RoutingContext.class).get(); <4>
+            return "show-first-page".equals(event.request().params().get("query-options"));
+        }
+
+        ...
+    }
+}
+----
+<1> Resource method `createOrUpdate` is only accessible by user with both `create` and `update` permissions.
+The `@PermissionsAllowed` is repeatable and all annotation instances are conjuncts.
+<2> User is granted access to `getItem` if he possess either `read` permission or `see` permission and one of actions (`all`, `detail`).
+Relation between permissions specified through `PermissionsAllowed#value` is disjunctive.
+<3> You can use any `java.security.Permission` implementation of your choice.
+By default, string-based permission is performed by the `io.quarkus.security.StringPermission`.
+<4> Permissions are not beans, therefore only way to obtain bean instances is programmatically via the `Arc.container()`.
+
+CAUTION: If you plan to use the `@PermissionsAllowed` on the IO thread, review the information in xref:security-proactive-authentication-concept.adoc[Proactive Authentication].
+NOTE: The `@PermissionsAllowed` is not repeatable on class-level due to limitations of Quarkus interceptors.
+Please find well-argued explanation in the xref:cdi-reference.adoc#repeatable-interceptor-bindings[Repeatable interceptor bindings] section of the CDI reference.
+
+On CDI beans, it is also possible to take custom `java.security.Permission` much further.
+Let's image that in your application, you have following service:
+
+[source,java]
+----
+import io.quarkus.security.PermissionsAllowed;
+import io.smallrye.mutiny.Uni;
+import jakarta.enterprise.context.ApplicationScoped;
+
+import java.security.Permission;
+import java.util.Objects;
+
+@ApplicationScoped
+public class GreetingsService {
+
+    @PermissionsAllowed(value = "ignored", permission = HelloPermission.class) <1>
```

Comment:
I'm not sure I understand what is happening here, use case, please clarify

Reply:
@sberyozkin 
Well, naturally if you are not sure, I'll have to clarify it in docs (extend example etc.). But till then, there is explanation:

```Java
   @PermissionsAllowed(value = "ignored", permission = HelloPermission.class, params = "greetings")
```
you say:

- permission name is "ignored" (that's a name, you can use "whatever" etc.)
- use HelloPermission class to perform permission checks
- use say method's formal parameter 'greetings' ( `public HelloPermission(String name, String[] actions, String greetings`) should be passed to the Permission constructor like this:
```Java
var svc = new GreetinsService();
svc.greetings(1, null, "hello");
var perm = new HelloPermission("ignored", null, "hello");
```
This way, you can allow access to the method based on values with which the method was invoked. It was suggested here by Stuart https://github.com/quarkusio/quarkus/issues/10988#issuecomment-902230311, and I remember using something similar in Spring (however they use AOP, imo our interceptors are more effective).


I also thought it was nicely explained here https://github.com/quarkusio/quarkus-security/blob/33ba33ab7113c36dca8e14fc4802ea92e65f3940/src/main/java/io/quarkus/security/PermissionsAllowed.java#L186

Please give me a hint which part wasn't clear from docs as I don't know how to extend example (which part is unclear). Thank you, that will help me.

Reply:
Sorry @sberyozkin , this answer actually belongs to the comment right below (where params are not autodetected).

Reply:
As for this specific example, there is one difference - you have single parameter with data type String and you can see `HelloPermission` constructor accepts a `String`, so you don't have to (but you still can) give Quarkus hint which parameter should be passed to the constructor.

Reply:
btw using `ignored` may seem weird, but `java.security.Permission` has required permission name, it felt weird to allow it to be null. It smells like a hack to me.

Reply:
As for use case of passing parameters of the invoked method to the security annotation - I don't have any statistics around the world, but in the company I used to work, I saw it countless times. In my eyes, it's pretty cool capability.

Reply:
I can see why it can be interesting to offload this parameter validation to some custom handler, but it feels to me the idea of using permissions is overloaded here, with `HelloPermission` example, what does it have to do with the authorization control, it looks like a parameter validator.

I'm not sure why the use of `PermissionsAllowed` is needed at all, given that here we don't care about the permission name,  where we put whatever name we want, just so that a custom permission class can verify this parameter without actually checking the permission name and action, instead of using the tools dedicated to the verification of parameters such as Hibernate Validation, or custom JAX-RS filter which can be bound to the given method.

Your example with the Workday evaluator is somewhat different there we check 3 things: permission name, action and an extra property so at least there I can imagine that this extra property check might represent an extra authorization check.

I think this is what you should highlight in docs that this extra property check, in addition to the regular permission allowed check related to the permission name/action, gives an extra authorization capability.

Having a HelloPermission example, with `any` permission name, is confusing IMHO. Please consider showing your workday test example here instead

Reply:
Hey Michal, I think I have an idea how to express my doubt more precisely, Permission is a function of some mapping from the authenticated identity,  and it can not work with this Hello Permission example, thanks

Reply:
@sberyozkin now I understand. I provided entirely new example, let see if it is better. I'll resolve this conversation as I think we should start from scratch.